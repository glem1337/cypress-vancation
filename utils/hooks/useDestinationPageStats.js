import { useSelector, useDispatch } from 'react-redux';

import { MAPBOX_FEATURE_TYPE } from 'constants/searchDestinations';
import {
  searchDestinationParamsSelector,
  searchDestinationFiltersSelector,
  desktopMapVisibilitySelector,
} from 'state/concepts/search-destinations/selectors';
import isPresent from 'utils/isPresent';
import {
  clearAllFilter as clearAllFilterAction,
  toggleDesktopMapVisibility as toggleDesktopMapVisibilityAction,
} from 'state/concepts/search-destinations/actions';
import { loadingSelector } from 'state/data/selectors';
import { fetchCampersEndpoint } from 'state/concepts/campervan-rental/endpoints';
import { campersSelector } from 'state/concepts/campervan-rental/selectors';

function useDestinationPageStats() {
  const dispatch = useDispatch();

  const campers = useSelector(campersSelector);

  const searchParams = useSelector(searchDestinationParamsSelector);

  const persistedFilters = useSelector(searchDestinationFiltersSelector);

  const isDesktopMapVisible = useSelector(desktopMapVisibilitySelector);

  const areCampersFetching = useSelector(
    state => loadingSelector(state, fetchCampersEndpoint.endpoint),
  );

  /**
   * Is any filter applied.
   */
  const isAnyFilterApplied = () => {
    const keys = Object.keys(persistedFilters);

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];

      if (
        key === 'uuid'
        || key === 'isMobileVisible'
      ) {
        // eslint-disable-next-line no-continue
        continue;
      }

      const value = persistedFilters[key];

      if (isPresent(value)) {
        return true;
      }
    }

    return false;
  };

  /**
   * Clear all filters.
   */
  const clearAllFilters = () => {
    dispatch(clearAllFilterAction());
  };

  /**
   * Toggle desktop map visibility
   */
  const toggleDesktopMapVisibility = () => {
    dispatch(toggleDesktopMapVisibilityAction());
  };

  return {
    isSearchResultsPage: searchParams?.location?.type === MAPBOX_FEATURE_TYPE,
    isAnyFilterApplied: isAnyFilterApplied(),
    clearAllFilters,
    isDesktopMapVisible,
    toggleDesktopMapVisibility,
    areCampersFetching,
    campers,
  };
}

export default useDestinationPageStats;
