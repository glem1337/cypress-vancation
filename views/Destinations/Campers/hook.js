import { useSelector, useDispatch } from 'react-redux';
import { v4 as uuid } from 'uuid';
import * as R from 'ramda';

import {
  campersSelector,
  campersPageSelector,
  campersTotalSelector,
} from 'state/concepts/campervan-rental/selectors';
import { fetchCampers } from 'state/concepts/campervan-rental/actions';
import { setActiveCamperId } from 'state/concepts/search-destinations/actions';
import { searchDestinationParamsSelector } from 'state/concepts/search-destinations/selectors';

export const DATA_TYPE = {
  CAMPERS_CHUNK: 'CampersChunk',
  DESCRIPTION: 'Description',
  FAN_FACTS: 'FunFacts',
  SINGLE_FACT: 'SingleFact',
};

function useContainer() {
  const dispatch = useDispatch();

  const campers = useSelector(campersSelector);
  const page = useSelector(campersPageSelector);
  const total = useSelector(campersTotalSelector);
  const searchParams = useSelector(searchDestinationParamsSelector);

  const description = R.path(['location', 'description'], searchParams);
  const factsTitle = R.path(['location', 'funFactsTitle'], searchParams);
  const facts = R.path(['location', 'funFacts'], searchParams);

  /**
   * On pagination change handler
   */
  const onPaginationChange = (newPage) => {
    dispatch(fetchCampers({
      page: newPage,
    }));
  };

  /**
   * On mouse leave.
   */
  const onMouseLeave = () => {
    if (campers.length > 0) {
      dispatch(setActiveCamperId(null));
    }
  };

  /**
   * Attach fun facts and description
   */
  const prepareCampersLayout = () => {
    let campersData = R.compose(
      R.map(item => ({
        id: uuid(),
        campers: item,
        type: DATA_TYPE.CAMPERS_CHUNK,
      })),
      R.splitEvery(2),
    )(campers);

    if (campers.length > 0 && campers.length < 4) {
      campersData = [
        ...campersData,
        {
          id: uuid(),
          description,
          type: DATA_TYPE.DESCRIPTION,
        },
        {
          id: uuid(),
          factsTitle,
          facts: facts || [],
          type: DATA_TYPE.FAN_FACTS,
        },
      ];
    }

    if (campers.length >= 4 && page === 1) {
      const DESCRIPTION_INDEX = 1;

      campersData = R.insert(
        DESCRIPTION_INDEX,
        {
          id: uuid(),
          description,
          type: DATA_TYPE.DESCRIPTION,
        },
        campersData,
      );

      let fanFactsInsertIndex = 1;
      for (let i = 0; i < R.defaultTo([], facts).length; i += 1) {
        fanFactsInsertIndex += 2;
        const fact = facts[i];

        campersData = R.insert(
          fanFactsInsertIndex,
          {
            ...fact,
            type: DATA_TYPE.SINGLE_FACT,
          },
          campersData,
        );
      }
    }

    return campersData;
  };

  return {
    campers,
    campersData: prepareCampersLayout(),
    page,
    perPage: 20,
    total,
    onPaginationChange,
    onMouseLeave,
    prepareCampersLayout,
  };
}

export default useContainer;
