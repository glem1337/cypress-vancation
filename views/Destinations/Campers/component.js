import React from 'react';
import { Pagination, Divider } from 'antd';
import { FormattedMessage } from 'react-intl';

import useDestinationPageStats from 'utils/hooks/useDestinationPageStats';
import OtherPopularDestinations from 'views/Destinations/OtherPopularDestinations';

import CollapsibleText from '../CollapsibleText';
import CamperCard from '../CamperCard';
import InnerHtmlField from '../InnerHtmlField';

import useContainer, { DATA_TYPE } from './hook';

const Campers = () => {
  const { isSearchResultsPage } = useDestinationPageStats();

  const {
    campers,
    page,
    perPage,
    total,
    onPaginationChange,
    onMouseLeave,
    campersData,
  } = useContainer();

  return (
    <>
      <div
        className="search-page__results-campers"
        onMouseLeave={onMouseLeave}
      >
        {isSearchResultsPage && campers?.length === 0 && (
          <OtherPopularDestinations />
        )}
        {isSearchResultsPage && (
          <div className="search-page__results-campers__grid">
            {campers?.map(camper => (
              <CamperCard
                key={camper.id}
                camper={camper}
              />
            ))}
          </div>
        )}
        {!isSearchResultsPage && campersData.map(item => {
          if (item.type === DATA_TYPE.DESCRIPTION) {
            return (
              <div className="mb-24" key={item.id}>
                <CollapsibleText>
                  <InnerHtmlField html={item.description} />
                </CollapsibleText>
              </div>
            );
          }

          if (item.type === DATA_TYPE.FAN_FACTS) {
            return (
              <div key={item.id}>
                <Divider className="mt-24 mb-24" />
                <h2 className="text-headline mb-24">{item.factsTitle}</h2>
                {item.facts.map(fact => (
                  <div className="search-page__results-fact mb-24" key={fact.id}>
                    <img
                      className="search-page__results-fact-icon"
                      src="/images/Fun Fact.svg"
                      alt=""
                    />
                    <InnerHtmlField html={fact.text} />
                  </div>
                ))}
              </div>
            );
          }

          if (item.type === DATA_TYPE.SINGLE_FACT) {
            return (
              <div className="search-page__results-fact mb-24" key={item.id}>
                <img
                  className="search-page__results-fact-icon"
                  src="/images/Fun Fact.svg"
                  alt=""
                />
                <InnerHtmlField html={item.text} />
              </div>
            );
          }

          return (
            <div className="search-page__results-campers__grid mb-24" key={item.id}>
              {item.campers.map(camper => (
                <CamperCard
                  key={camper.id}
                  camper={camper}
                />
              ))}
            </div>
          );
        })}
      </div>
      {total > perPage && (
        <div className="d-flex flex-column align-items-center mt-24 mt-md-40">
          <Pagination
            current={page}
            total={total}
            pageSize={perPage}
            showSizeChanger={false}
            onChange={onPaginationChange}
          />
          <div className="mt-24">
            <FormattedMessage
              id="search-destination.paginationPhrase"
              values={{
                page,
                perPage: Math.min(perPage, total),
                total,
              }}
            />
          </div>
        </div>
      )}
    </>
  );
};

export default Campers;
