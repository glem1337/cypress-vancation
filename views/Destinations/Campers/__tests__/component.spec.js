import React from 'react';
import { shallow } from 'enzyme';

import useDestinationPageStats from 'utils/hooks/useDestinationPageStats';

import Campers from '../component';
import useContainer, { DATA_TYPE } from '../hook';

const mockedUseDestinationPageStats = {
  isSearchResultsPage: false,
};
jest.mock('utils/hooks/useDestinationPageStats', () => jest.fn(() => mockedUseDestinationPageStats));

const mockedHookData = {
  campers: [{ id: 1 }, { id: 2 }],
  page: 1,
  perPage: 20,
  total: 200,
  onPaginationChange: jest.fn(),
  onMouseLeave: jest.fn(),
  campersData: [
    {
      id: 1,
      type: DATA_TYPE.DESCRIPTION,
      description: 'test description',
    },
    {
      id: 2,
      type: DATA_TYPE.FAN_FACTS,
      factsTitle: 'test title',
      facts: [{ id: 11, text: 'test text' }],
    },
    {
      id: 3,
      type: DATA_TYPE.SINGLE_FACT,
      text: 'test text',
    },
    {
      id: 4,
      type: DATA_TYPE.CAMPERS_CHUNK,
      campers: [{ id: 11 }],
    },
  ],
};
jest.mock('../hook', () => ({
  __esModule: true,
  ...jest.requireActual('../hook'),
  default: jest.fn(() => mockedHookData),
}));

describe('Campers component tests', () => {
  it('matches snapshot', () => {
    const component = shallow(<Campers />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when there are no campers', () => {
    useContainer.mockReturnValueOnce({
      ...mockedHookData,
      campers: null,
    });

    const component = shallow(<Campers />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when isSearchResultsPage', () => {
    useDestinationPageStats.mockReturnValueOnce({ isSearchResultsPage: true });

    const component = shallow(<Campers />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when isSearchResultsPage and there are no campers', () => {
    useDestinationPageStats.mockReturnValueOnce({ isSearchResultsPage: true });
    useContainer.mockReturnValueOnce({
      ...mockedHookData,
      campers: null,
    });

    const component = shallow(<Campers />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when isSearchResultsPage and there are campers', () => {
    useDestinationPageStats.mockReturnValueOnce({ isSearchResultsPage: true });
    useContainer.mockReturnValueOnce({
      ...mockedHookData,
      campers: [{ id: 1 }],
    });

    const component = shallow(<Campers />);

    expect(component).toMatchSnapshot();
  });
});
