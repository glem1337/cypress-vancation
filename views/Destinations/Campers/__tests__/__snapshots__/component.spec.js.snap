// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campers component tests matches snapshot 1`] = `
<Fragment>
  <div
    className="search-page__results-campers"
    onMouseLeave={[MockFunction]}
  >
    <div
      className="mb-24"
      key="1"
    >
      <CollapsibleText>
        <InnerHtmlField
          WrapperTag="span"
          html="test description"
        />
      </CollapsibleText>
    </div>
    <div
      key="2"
    >
      <Divider
        className="mt-24 mb-24"
      />
      <h2
        className="text-headline mb-24"
      >
        test title
      </h2>
      <div
        className="search-page__results-fact mb-24"
        key="11"
      >
        <img
          alt=""
          className="search-page__results-fact-icon"
          src="/images/Fun Fact.svg"
        />
        <InnerHtmlField
          WrapperTag="span"
          html="test text"
        />
      </div>
    </div>
    <div
      className="search-page__results-fact mb-24"
      key="3"
    >
      <img
        alt=""
        className="search-page__results-fact-icon"
        src="/images/Fun Fact.svg"
      />
      <InnerHtmlField
        WrapperTag="span"
        html="test text"
      />
    </div>
    <div
      className="search-page__results-campers__grid mb-24"
      key="4"
    >
      <CamperCard
        camper={
          Object {
            "id": 11,
          }
        }
        key="11"
      />
    </div>
  </div>
  <div
    className="d-flex flex-column align-items-center mt-24 mt-md-40"
  >
    <Pagination
      current={1}
      onChange={[MockFunction]}
      pageSize={20}
      showSizeChanger={false}
      total={200}
    />
    <div
      className="mt-24"
    >
      <Memo(FormattedMessage)
        id="search-destination.paginationPhrase"
        values={
          Object {
            "page": 1,
            "perPage": 20,
            "total": 200,
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Campers component tests matches snapshot when isSearchResultsPage 1`] = `
<Fragment>
  <div
    className="search-page__results-campers"
    onMouseLeave={[MockFunction]}
  >
    <div
      className="search-page__results-campers__grid"
    >
      <CamperCard
        camper={
          Object {
            "id": 1,
          }
        }
        key="1"
      />
      <CamperCard
        camper={
          Object {
            "id": 2,
          }
        }
        key="2"
      />
    </div>
  </div>
  <div
    className="d-flex flex-column align-items-center mt-24 mt-md-40"
  >
    <Pagination
      current={1}
      onChange={[MockFunction]}
      pageSize={20}
      showSizeChanger={false}
      total={200}
    />
    <div
      className="mt-24"
    >
      <Memo(FormattedMessage)
        id="search-destination.paginationPhrase"
        values={
          Object {
            "page": 1,
            "perPage": 20,
            "total": 200,
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Campers component tests matches snapshot when isSearchResultsPage and there are campers 1`] = `
<Fragment>
  <div
    className="search-page__results-campers"
    onMouseLeave={[MockFunction]}
  >
    <div
      className="search-page__results-campers__grid"
    >
      <CamperCard
        camper={
          Object {
            "id": 1,
          }
        }
        key="1"
      />
    </div>
  </div>
  <div
    className="d-flex flex-column align-items-center mt-24 mt-md-40"
  >
    <Pagination
      current={1}
      onChange={[MockFunction]}
      pageSize={20}
      showSizeChanger={false}
      total={200}
    />
    <div
      className="mt-24"
    >
      <Memo(FormattedMessage)
        id="search-destination.paginationPhrase"
        values={
          Object {
            "page": 1,
            "perPage": 20,
            "total": 200,
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Campers component tests matches snapshot when isSearchResultsPage and there are no campers 1`] = `
<Fragment>
  <div
    className="search-page__results-campers"
    onMouseLeave={[MockFunction]}
  >
    <div
      className="search-page__results-campers__grid"
    />
  </div>
  <div
    className="d-flex flex-column align-items-center mt-24 mt-md-40"
  >
    <Pagination
      current={1}
      onChange={[MockFunction]}
      pageSize={20}
      showSizeChanger={false}
      total={200}
    />
    <div
      className="mt-24"
    >
      <Memo(FormattedMessage)
        id="search-destination.paginationPhrase"
        values={
          Object {
            "page": 1,
            "perPage": 20,
            "total": 200,
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`Campers component tests matches snapshot when there are no campers 1`] = `
<Fragment>
  <div
    className="search-page__results-campers"
    onMouseLeave={[MockFunction]}
  >
    <div
      className="mb-24"
      key="1"
    >
      <CollapsibleText>
        <InnerHtmlField
          WrapperTag="span"
          html="test description"
        />
      </CollapsibleText>
    </div>
    <div
      key="2"
    >
      <Divider
        className="mt-24 mb-24"
      />
      <h2
        className="text-headline mb-24"
      >
        test title
      </h2>
      <div
        className="search-page__results-fact mb-24"
        key="11"
      >
        <img
          alt=""
          className="search-page__results-fact-icon"
          src="/images/Fun Fact.svg"
        />
        <InnerHtmlField
          WrapperTag="span"
          html="test text"
        />
      </div>
    </div>
    <div
      className="search-page__results-fact mb-24"
      key="3"
    >
      <img
        alt=""
        className="search-page__results-fact-icon"
        src="/images/Fun Fact.svg"
      />
      <InnerHtmlField
        WrapperTag="span"
        html="test text"
      />
    </div>
    <div
      className="search-page__results-campers__grid mb-24"
      key="4"
    >
      <CamperCard
        camper={
          Object {
            "id": 11,
          }
        }
        key="11"
      />
    </div>
  </div>
  <div
    className="d-flex flex-column align-items-center mt-24 mt-md-40"
  >
    <Pagination
      current={1}
      onChange={[MockFunction]}
      pageSize={20}
      showSizeChanger={false}
      total={200}
    />
    <div
      className="mt-24"
    >
      <Memo(FormattedMessage)
        id="search-destination.paginationPhrase"
        values={
          Object {
            "page": 1,
            "perPage": 20,
            "total": 200,
          }
        }
      />
    </div>
  </div>
</Fragment>
`;
